modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/SnapshotService.scala

```diff
@@ -27,48 +28,59 @@ object SnapshotService {
 
   def make[F[_]: Async](
     snapshotStorage: SnapshotStorage[F],
-    blocksQueue: Queue[F, Set[Signed[DAGBlock]]],
+    snapshotPreconditionsValidator: SnapshotPreconditionsValidator[F],
+    l1DataQueue: Queue[F, L1Data],
     config: SnapshotConfig
   ): SnapshotService[F] =
     new SnapshotService[F] {
 
-      val snapshotCreation: Stream[F, GlobalSnapshot] =
+      val pipeline: Stream[F, Either[Signed[DAGBlock], SnapshotReference]] =
         Stream
-          .fromQueueUnterminated(blocksQueue)
-          .evalMap(createSnapshot)
-          .through(switchRepeat(config.fallbackTriggerTimeout, Stream.eval(createSnapshot())))
+          .fromQueueUnterminated(l1DataQueue)
+          .zip(Stream.eval(snapshotStorage.getLastSnapshotHeight))
+          .flatMap {
+            case (data, lastSnapshotHeight) =>
+              snapshotPreconditionsValidator.validate(lastSnapshotHeight, data) match {
+                case Valid(lastSnapshotHeight) =>
+                  Stream.emit(data.block.asLeft[SnapshotReference]) ++ Stream
```

Comment:
I think it should be reversed, first the block and then the `SnapshotReference`.

Reply:
I just realized that it is already reversed, I just misread that fragment 🤦